File: frontend/js/engine/lexer.js
export class EngineError extends Error {
  constructor(code, message) {
    super(message);
    this.code = code;
  }
}

export function tokenize(source) {
  const tokens = [];
  let i = 0;
  const isDigit = (c) => c >= '0' && c <= '9';
  const isIdentStart = (c) => /[A-Za-z_\u03c0]/.test(c);
  const isIdentPart = (c) => /[A-Za-z0-9_]/.test(c);

  while (i < source.length) {
    const c = source[i];
    if (/\s/.test(c)) {
      i++;
      continue;
    }

    if (isDigit(c) || c === '.') {
      let start = i;
      let dotCount = 0;
      let sawDigit = false;
      while (i < source.length && (isDigit(source[i]) || source[i] === '.')) {
        if (source[i] === '.') dotCount++;
        else sawDigit = true;
        i++;
      }
      if (dotCount > 1 || !sawDigit) {
        throw new EngineError('SYNTAX_ERROR', `Malformed number at position ${start}`);
      }
      if (i < source.length && (source[i] === 'e' || source[i] === 'E')) {
        let look = i + 1;
        if (look < source.length && (source[look] === '+' || source[look] === '-')) look++;
        if (look < source.length && isDigit(source[look])) {
          i++;
          if (source[i] === '+' || source[i] === '-') i++;
          while (i < source.length && isDigit(source[i])) i++;
        }
      }

      const numStr = source.slice(start, i);

      if (i < source.length && source[i] === '[') {
        const unitStart = i + 1;
        const closeIdx = source.indexOf(']', unitStart);
        if (closeIdx === -1) {
          throw new EngineError('SYNTAX_ERROR', `Unclosed unit bracket at position ${i}`);
        }
        const unitStr = source.slice(unitStart, closeIdx).trim();
        if (unitStr.length === 0) {
          throw new EngineError('SYNTAX_ERROR', `Empty unit brackets at position ${i}`);
        }
        tokens.push({
          type: 'NUMBER_WITH_UNIT',
          text: source.slice(start, closeIdx + 1),
          value: numStr,
          unit: unitStr,
        });
        i = closeIdx + 1;
        continue;
      }

      tokens.push({
        type: 'NUMBER',
        text: numStr,
        value: numStr,
      });
      continue;
    }

    if (isIdentStart(c)) {
      let start = i;
      while (i < source.length && isIdentPart(source[i])) i++;
      tokens.push({ type: 'IDENTIFIER', text: source.slice(start, i) });
      continue;
    }

    const single = {
      '+': 'PLUS',
      '-': 'MINUS',
      '\u2212': 'MINUS',
      '*': 'STAR',
      '\u00D7': 'STAR',
      '/': 'SLASH',
      '\u00F7': 'SLASH',
      '%': 'PERCENT',
      '^': 'CARET',
      '!': 'BANG',
      '(': 'LPAREN',
      ')': 'RPAREN',
      ',': 'COMMA',
      '=': 'EQUALS',
    };

    if (single[c]) {
      tokens.push({ type: single[c], text: c });
      i++;
      continue;
    }

    throw new EngineError('SYNTAX_ERROR', `Unexpected character '${c}' at position ${i}`);
  }

  tokens.push({ type: 'EOF', text: '' });
  return tokens;
}

export function validateUnitTokenization(expr = 'mass = 50[kg] * 9.81[m/s^2]') {
  const tokens = tokenize(expr);
  const unitsOnly = tokens.filter((t) => t.type === 'NUMBER_WITH_UNIT');
  return {
    input: expr,
    tokens,
    extractedUnits: unitsOnly.map((t) => ({ value: t.value, unit: t.unit })),
    valid:
      unitsOnly.length === 2 &&
      unitsOnly[0].value === '50' &&
      unitsOnly[0].unit === 'kg' &&
      unitsOnly[1].value === '9.81' &&
      unitsOnly[1].unit === 'm/s^2',
  };
}

File: backend/src/main/java/com/calcforge/engine/Lexer.java
package com.calcforge.engine;

import com.calcforge.engine.ExpressionException.ErrorCode;

import java.util.ArrayList;
import java.util.List;

/**
 * Turns a raw expression string into a flat list of {@link Token}s.
 *
 * <p>Supports standard ASCII operators plus a handful of Unicode aliases that
 * calculator keypads commonly produce ({@code × ÷ −}), so pasted or keypad-built
 * input behaves identically to typed ASCII input.</p>
 */
public class Lexer {

    private final String source;
    private int pos = 0;

    public Lexer(String source) {
        this.source = source == null ? "" : source;
    }

    public List<Token> tokenize() {
        List<Token> tokens = new ArrayList<>();
        while (true) {
            skipWhitespace();
            if (isAtEnd()) {
                tokens.add(new Token(TokenType.EOF, "", pos));
                break;
            }

            char c = peek();
            int start = pos;

            if (Character.isDigit(c) || c == '.') {
                tokens.add(readNumber());
                continue;
            }

            if (isIdentifierStart(c)) {
                tokens.add(readIdentifier());
                continue;
            }

            switch (c) {
                case '+' -> {
                    advance();
                    tokens.add(new Token(TokenType.PLUS, "+", start));
                }
                case '-', '\u2212' -> { // ASCII hyphen-minus or Unicode minus sign
                    advance();
                    tokens.add(new Token(TokenType.MINUS, "-", start));
                }
                case '*', '\u00D7' -> { // '*' or Unicode multiplication sign ×
                    advance();
                    tokens.add(new Token(TokenType.STAR, "*", start));
                }
                case '/', '\u00F7' -> { // '/' or Unicode division sign ÷
                    advance();
                    tokens.add(new Token(TokenType.SLASH, "/", start));
                }
                case '%' -> {
                    advance();
                    tokens.add(new Token(TokenType.PERCENT, "%", start));
                }
                case '^' -> {
                    advance();
                    tokens.add(new Token(TokenType.CARET, "^", start));
                }
                case '!' -> {
                    advance();
                    tokens.add(new Token(TokenType.BANG, "!", start));
                }
                case '(' -> {
                    advance();
                    tokens.add(new Token(TokenType.LPAREN, "(", start));
                }
                case ')' -> {
                    advance();
                    tokens.add(new Token(TokenType.RPAREN, ")", start));
                }
                case ',' -> {
                    advance();
                    tokens.add(new Token(TokenType.COMMA, ",", start));
                }
                default -> throw new ExpressionException(
                        ErrorCode.SYNTAX_ERROR,
                        "Unexpected character '" + c + "' at position " + pos);
            }
        }
        return tokens;
    }

    private Token readNumber() {
        int start = pos;
        int dotCount = 0;
        boolean sawDigit = false;

        while (!isAtEnd() && (Character.isDigit(peek()) || peek() == '.')) {
            if (peek() == '.') {
                dotCount++;
            } else {
                sawDigit = true;
            }
            advance();
        }

        if (dotCount > 1) {
            throw new ExpressionException(ErrorCode.SYNTAX_ERROR,
                    "Malformed number with multiple decimal points at position " + start);
        }

        if (!sawDigit) {
            throw new ExpressionException(ErrorCode.SYNTAX_ERROR,
                    "Malformed number at position " + start);
        }

        // Optional exponent: e / E, optional sign, one or more digits.
        if (!isAtEnd() && (peek() == 'e' || peek() == 'E')) {
            int savedPos = pos;
            int lookahead = pos + 1;
            if (lookahead < source.length() && (source.charAt(lookahead) == '+' || source.charAt(lookahead) == '-')) {
                lookahead++;
            }
            if (lookahead < source.length() && Character.isDigit(source.charAt(lookahead))) {
                advance(); // consume e/E
                if (peek() == '+' || peek() == '-') {
                    advance();
                }
                while (!isAtEnd() && Character.isDigit(peek())) {
                    advance();
                }
            } else {
                pos = savedPos; // not actually an exponent (e.g. "3e" used as "3*e")
            }
        }

        if (!isAtEnd() && peek() == '[') {
            int closeBracket = source.indexOf(']', pos);
            if (closeBracket == -1) {
                throw new ExpressionException(ErrorCode.SYNTAX_ERROR,
                        "Unclosed unit bracket '[' at position " + pos);
            }
            pos = closeBracket + 1;
            return new Token(TokenType.NUMBER_WITH_UNIT, source.substring(start, pos), start);
        }

        return new Token(TokenType.NUMBER, source.substring(start, pos), start);
    }

    private Token readIdentifier() {
        int start = pos;
        while (!isAtEnd() && isIdentifierPart(peek())) {
            advance();
        }
        return new Token(TokenType.IDENTIFIER, source.substring(start, pos), start);
    }

    private boolean isIdentifierStart(char c) {
        return Character.isLetter(c) || c == '_' || c == '\u03C0'; // allow literal 'π'
    }

    private boolean isIdentifierPart(char c) {
        return Character.isLetterOrDigit(c) || c == '_';
    }

    private void skipWhitespace() {
        while (!isAtEnd() && Character.isWhitespace(peek())) {
            advance();
        }
    }

    private boolean isAtEnd() {
        return pos >= source.length();
    }

    private char peek() {
        return source.charAt(pos);
    }

    private void advance() {
        pos++;
    }
}
